## ANLP Assignment 8

In [12]:
import numpy
import spacy
from math import*
import pandas as pd
nlp = spacy.load('en')

## Task 1 

In [13]:
Ex_txt_raw = open("ex1.txt")
Ex_txt_processed = Ex_txt_raw.read()
Ex_txt_processed2=nlp(Ex_txt_processed)

comp = nlp(u"My poodle likes food in general.")

for sent in Ex_txt_processed2.sents:
    x=sent.text
    y=nlp(x)
    print("The similarity score for the sentance '",x, "' aganist the seed sentance is is ", y.similarity(comp))

The similarity score for the sentance ' This is a sentence. ' aganist the seed sentance is is  0.5543370457325273
The similarity score for the sentance ' This is another sentence. ' aganist the seed sentance is is  0.5430651457911527
The similarity score for the sentance ' John likes milk. ' aganist the seed sentance is is  0.5666850379747415
The similarity score for the sentance ' Peter eats candy. ' aganist the seed sentance is is  0.5452489053230556
The similarity score for the sentance ' My dog likes bones.
 ' aganist the seed sentance is is  0.7247384874616155


Based on the results above we can conclude that "My dog likes bones" is the sentance that is most similar to "My poodle likes food in general"

## Task 2 

### a) Most similar words

In [14]:
tokens1 = nlp(u'John loves apples')
tokens2 = nlp(u'some man likes fruit')

Euclidian_dist=[]
for token1 in tokens1:
    for token2 in tokens2:
        t1 = token1.vector
        ##print(t1)
        t2 = token2.vector
        ##print(t2)
        simi=token1.similarity(token2)
        val=sqrt(sum(pow(x-y,2) for x, y in zip(t1, t2)))
        Euclidian_dist.append([simi,val])
        print("The similarity between",token1.text,"and", token2.text, "is",token1.similarity(token2))
        
        

The similarity between John and some is -0.014235469
The similarity between John and man is 0.26821983
The similarity between John and likes is -0.08188061
The similarity between John and fruit is 0.2018577
The similarity between loves and some is 0.0007409384
The similarity between loves and man is 0.1280556
The similarity between loves and likes is 0.46491706
The similarity between loves and fruit is 0.10413371
The similarity between apples and some is -0.03522985
The similarity between apples and man is 0.16096427
The similarity between apples and likes is 0.10055058
The similarity between apples and fruit is 0.3141492


Based on the above results we can say that loves and likes are the most similar words

### b)Python code that computes the distance between them.

In [15]:
Euclidian_dist=pd.DataFrame(Euclidian_dist)
Euclidian_dist.columns=['similarity','distance']
print(Euclidian_dist.loc[Euclidian_dist['similarity'].idxmax(),'distance'])

28.13667312321284


The Euclidian distance between the most similar words is 28.13667312321284

### c)Sum up and average over the distances to print out the Word Mover's Distance between the two lines

In [16]:
Euclidian_dist['distance'].mean()

34.244884795762836

### Task 3

### Steps
1)Read the necessary test 
2)Split by sentance using spacy 
3)Use spacy to identify the Subject, verb and object from the sentence 
4)Write the results in a text file

In [17]:
Ex_txt_raw = open("data.txt")
Ex_txt_processed = Ex_txt_raw.read()
Ex_txt_processed2=nlp(Ex_txt_processed)
result=''
for sent in Ex_txt_processed2.sents:
    x=sent.text
    token = nlp(x)
    subject_res=[]
    object_res=[]
    verb_res=[]
    for k in token:
        if(k.dep_ == "nsubj"):
            subject_res.append(k)
        if(k.dep_ == "dobj"):
            object_res.append(k)
        if(k.pos_ == "VERB" and k.dep_ != "aux"):
            verb_res.append(k)
    y1="For the sentance :"  + str(x) +'\n'
    y2="The Subject is "+str(subject_res) +'\n'
    y3="The Verb is "+str(verb_res) +'\n'
    y4="The Object is "+str(object_res) +'\n'
    res=y1+y2+y3+y4
    print(res)
    result=result+res
tex_res = open("Final_Output.txt", "w")
tex_res.write(str(result))
tex_res.close()

For the sentance :Science and mathematics courses are more beneficial to students than literature and history classes for three reasons.
The Subject is [courses]
The Verb is [are]
The Object is []

For the sentance :First, the knowledge of science and math is essential to learning other important disciplines.
The Subject is [knowledge]
The Verb is [is, learning]
The Object is [disciplines]

For the sentance :Second, it is more important in everyday life than literature and history.
The Subject is [it]
The Verb is [is]
The Object is []

For the sentance :And finally, science and math courses develop logic skills the way the humanities courses cant.


The Subject is [courses, courses]
The Verb is [develop, ca]
The Object is [skills]

For the sentance :First and foremost importantly, science and mathematics are essential for learning economics, finance, or mechanics, which are disciplines that are invaluable in todays marketplace.
The Subject is [science, which, that]
The Verb is [are, le

## Reference

https://spacy.io/usage/vectors-similarity